sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2983,6 +2983,9 @@ case class Sequence(
 
   override def nullable: Boolean = children.exists(_.nullable)
 
+  // If step is defined, then an error will be thrown if the start and stop do not satisfy the step.
+  override lazy val throwable: Boolean = stepOpt.isDefined
```

Comment:
Can we leverage `NoThrow` trait?

Reply:
the problem with `NoThrow` is that it doesn't mark that many expressions right now (i.e. if we only push down `NoThrow` expressions then we would be not pushing down most things). we want the opposite here (marking only certain expressions not to push down). defining it this way also allows us to be more precise with defining what is throwable, compared to just setting it for the class

Reply:
These two are similar but be conservative in two opposite directions. Filter pushdown is a very important feature and we don't want to suddenly disable it for many predicates. We only mark `sequence` as throwable for now.

`NoThrow` was added for a new optimization and only a few expressions are marked as `NoThrow` that can be optimized.

Since runtime error is rare, I think the current solution is better to produce more optimized plans for most cases.

Reply:
If it's rare, why do we add a property to `Expression`?

Reply:
Because it's indeed an expression property. `NoThrow` is not the right way to add properties as the caller side needs to do recursion: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L53-L59 , which is not good for code reuse (you need to call a util function)

Reply:
Should probably remove `NoThrow` then ... I approved it for now.